cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+* CID is created or deleted - all Agents regenerate all of their endpoints.
+* Network policies are changed.
+
+IPCache is updated when remote CEPs/CESs are received.
+
+## Label changes
+
+
+### Pod
+
+Pod/Deployment label changes result in pod recreation, which follows the same pod creation path previously described.
+
+
+### Namespace
+
+When namespace labels change, Operator and Agents get the update at the same time. Agents wait until Operator processes the namespace label change update and creates all the new CIDs. Agents keep retrying to match CIDs for all their local endpoints until they are received through the CID watcher.
+
+
+## Simultaneous CID management compatibility
+
+The only problem with simultaneous CID management (both Operator and Agents are managing CIDs) is when namespace labels are changed. The only bad thing that happens is that there is CID duplication. This is an existing problem already, that doesn't cause disruption.
+
+Simultaneous CID management is only expected to happen during upgrade from Agents managing CIDs to Operator managing CIDs.
+
+
+## CID deduplication
+
+Operator will not create duplicate CIDs, but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
+
+## KVStore compatibility
+
+[KVStore](https://docs.cilium.io/en/stable/kvstore/) is a key-value store that is mostly used for sharing Cilium mappings between IPs, identities (IPCache) and labels, across the cluster ([what's inside Cilium’s KVStore](http://arthurchiao.art/blog/whats-inside-cilium-etcd/)).
+
+Operator managing CIDs is intended to work with both `identity-allocation-mode=kvstore` and `identity-allocation-mode=crd`. CRD being the default, the initial implementation will use the CRD allocation mode, and later expanded to other allocation modes.
+
+All features that rely on using kvstore will not work until CID controller is expanded to work with kvstore allocation mode. The work to enable kvstore compatibility is estimated to of medium size and it will be tracked in a separate GitHub issue (https://github.com/cilium/cilium/issues/34865).
+
+
+# Enablement
+
+The feature will be guarded by a flag `operator-manages-identities`, that can be set in container arguments in Operator’s and Agent’s manifests. The default values for both will be `false`, meaning that Operator doesn’t manage CIDs, while Agents do.
+
+
+## Long term plan
+
+The long term plan is to make Operator managing CIDs the default, and deprecate the part of Agent that is currently used to manage CIDs.
+
+
+# Migration
+
+Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for pod creation. The control plane side of network policies and other features relying on CIDs will not carry out any required changes during this time, causing them to stop working.
```

Comment:
nit: I'm struggling a bit to understand what the second half of this paragraph is stating. Is there something more than "You can't successfully create Pods while the Operator is down"?

Reply:
Pod creation is the main one, but there are others. I'm not sure if I should list them here, or mentioning that pod creation will fail is sufficient.
- namespace label changes
- changing identity relevant labels filter
- updating deployment labels (it might not be obvious that this in fact redeploy pods)

Reply:
I see yeah, it gets pretty nuanced. I would rather highlight these specific aspects. It'd also be better to avoid wording like `causing them to stop working`, as this implies a failure of some kind but is not very specific about the failure condition. 

Maybe something like,
```suggestion
Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for assigning new Security Identities. Either of these components must be available and healthy in order to:
- Deploy workloads with a new set of identity-relevant labels
- Update existing workloads with different identity-relevant labels
- Modify the identity-relevant labels filter
- Modify namespace labels and enforce policies for the newly changed labels
```

Reply:
Something else I had in mind at the same time was that if a new instance of an existing deployment is created or a new Pod is created where the Security Identity is already allocated, then it should actually be possible to deploy that Pod without these components running.

Reply:
Yes, you're right. It would be possible to deploy new pods that have labels that match an existing CID.

The blast radius is much clearer the way you phrased it.